util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+	ret = fi_recv(ct->ep, ct->rx_buf, ct->rx_size, fi_mr_desc(ct->mr), 0, &ct->rx_ctx);
+
+	pp_banner_fabric_info(ct);
+
+	sizes_cnt = generate_test_sizes(&ct->opts, ct->fi->ep_attr->max_msg_size, &sizes, PP_SIZE_MAX_POWER_TWO);
+
+	PP_DEBUG("Count of sizes to test : %d\n", sizes_cnt);
+
+	for (i = 0; i < sizes_cnt; i++) {
+		ct->opts.transfer_size = sizes[i];
+		if (ct->opts.transfer_size > ct->fi->ep_attr->max_msg_size) {
+			PP_DEBUG("Transfer size too high for endpoint : %d\n", ct->opts.transfer_size);
+			continue;
+		}
+		init_test(ct, &(ct->opts), ct->test_name, sizeof(ct->test_name));
+		ret = pingpong(ct);
+		if (ret)
+			return ret;
+	}
+
+	free(sizes);
+
+	return pp_finalize(ct);
+}
+
+static int run_pingpong_rdm(struct ct_pingpong *ct)
+{
+	int i, ret, sizes_cnt;
+	int *sizes = NULL;
+
+	PP_DEBUG("Selected endpoint : RDM\n");
+
+	ret = pp_init_fabric(ct);
+	if (ret)
+		return ret;
+
+	pp_banner_fabric_info(ct);
+
+	sizes_cnt = generate_test_sizes(&ct->opts, ct->fi->ep_attr->max_msg_size, &sizes, PP_SIZE_MAX_POWER_TWO);
+
+	PP_DEBUG("Count of sizes to test : %d\n", sizes_cnt);
+
+	for (i = 0; i < sizes_cnt; i++) {
+		ct->opts.transfer_size = sizes[i];
+		if (ct->opts.transfer_size > ct->fi->ep_attr->max_msg_size) {
+			PP_DEBUG("Transfer size too high for endpoint : %d\n", ct->opts.transfer_size);
+			continue;
+		}
+		init_test(ct, &(ct->opts), ct->test_name, sizeof(ct->test_name));
+		ret = pingpong(ct);
+		if (ret)
+			return ret;
+	}
+
+	free(sizes);
+
+	return pp_finalize(ct);
+}
+
+static int run_pingpong_msg(struct ct_pingpong *ct)
+{
+	int i, ret, sizes_cnt;
+	int *sizes = NULL;
+
+	PP_DEBUG("Selected endpoint : DGRAM\n");
+
+	ret = pp_ctrl_init(ct);
+	if (ret) {
+		return ret;
+	}
+
+	ret = pp_ctrl_txrx_data_port(ct);
+	if (ret) {
+		return ret;
+	}
+
+	if (!ct->opts.dst_addr) {
+		ret = pp_start_server(ct);
+		if (ret)
+			return ret;
+	}
+
+	ret = ct->opts.dst_addr ? pp_client_connect(ct) : pp_server_connect(ct);
+	if (ret) {
+		return ret;
+	}
+
+	pp_banner_fabric_info(ct);
+
+	sizes_cnt = generate_test_sizes(&ct->opts, ct->fi->ep_attr->max_msg_size, &sizes, PP_SIZE_MAX_POWER_TWO);
+
+	PP_DEBUG("Count of sizes to test : %d\n", sizes_cnt);
+
+	for (i = 0; i < sizes_cnt; i++) {
+		ct->opts.transfer_size = sizes[i];
+		if (ct->opts.transfer_size > ct->fi->ep_attr->max_msg_size) {
+			PP_DEBUG("Transfer size too high for endpoint : %d\n", ct->opts.transfer_size);
+			continue;
+		}
+		init_test(ct, &(ct->opts), ct->test_name, sizeof(ct->test_name));
+		ret = pingpong(ct);
+		if (ret)
+			goto out;
+	}
+
+	free(sizes);
+
+	ret = pp_finalize(ct);
+out:
+	fi_shutdown(ct->ep, 0);
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int ret, op;
+
+	ret = EXIT_SUCCESS;
+
+	struct ct_pingpong ct;
+
+	pp_init_ct_pingpong(&ct);
+	ct.opts = (struct pp_opts) {
+		.options = 0,
+		.iterations = 1000,
+		.transfer_size = 1024,
+		.sizes_enabled = PP_DEFAULT_SIZE,
+		.argc = argc, .argv = argv
```

Comment:
> can we add arguments to override the mode bits?

Did you have specific mode bits in mind? I think the original plan was this test was to keep it simple and only support `FI_LOCAL_MR`/`FI_CONTEXT`.